### PR TITLE
CLOUDP-222249: Deployment with translation layer and Atlas SDK

### PR DESCRIPTION
**Pull request by helderjs**
### All Submissions:

* [x] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if there is one).
* [ ] Update docs/release-notes/release-notes-template.md if your changes should be included in the release notes for the next release.

### CODE REVIEW

File: internal/pointer/pointer.go [hunk 1]
```diff
@@ -23,6 +23,22 @@ func GetOrDefault[T any](ptr *T, defaultValue T) T {
 	return defaultValue
 }
 
+// GetOrPointerToDefault returns the value of a pointer or a pointer to default value
+func GetOrPointerToDefault[T any](ptr *T, defaultValue T) *T {
```
Comment (josvazg) [inline internal/pointer/pointer.go:27]:
```suggestion
// GetOrDefaultPointer returns the value of a pointer or a pointer to default value
func GetOrDefaultPointer[T any](ptr *T, defaultValue T) *T {
```

File: pkg/controller/atlasdeployment/atlasdeployment_controller.go [hunk 2]
```diff
@@ -111,181 +107,123 @@ func (r *AtlasDeploymentReconciler) Reconcile(context context.Context, req ctrl.
...
+	atlasSdkClient, _, err := r.AtlasProvider.SdkClient(ctx.Context, project.ConnectionSecretObjectKey(), log)
 	if err != nil {
 		result := workflow.Terminate(workflow.AtlasAPIAccessNotConfigured, err.Error())
-		workflowCtx.SetConditionFromResult(api.DeploymentReadyType, result)
-		return result.ReconcileResult(), nil
-	}
-	workflowCtx.SdkClient = atlasSdkClient
-
-	// Allow users to specify M0/M2/M5 deployments without providing TENANT for Normal and Serverless deployments
-	r.verifyNonTenantCase(deployment)
```
Comment (igor-karpukhin) [inline pkg/controller/atlasdeployment/atlasdeployment_controller.go:165]: Are we not going to support that anymore?

**Review by igor-karpukhin — COMMENTED** [review]
LGTM in general. Massive work here 👍 One question about the M0/M2/M5. See my comment above

File: pkg/controller/atlasdeployment/atlasdeployment_controller.go (hunk 2 above)
Comment (igor-karpukhin) [reply pkg/controller/atlasdeployment/atlasdeployment_controller.go:165]: After a short discussion with @helderjs we decided to leave this as is and create a follow up PR to resolve this issue

**Review by igor-karpukhin — APPROVED** [review]

File: pkg/controller/atlasdeployment/advanced_deployment.go [hunk 3]
```diff
@@ -1,333 +1,128 @@
...
 	databaseUsers := akov2.AtlasDatabaseUserList{}
 	err := r.Client.List(ctx.Context, &databaseUsers, &client.ListOptions{})
 	if err != nil {
-		return workflow.Terminate(workflow.Internal, err.Error())
+		return err
 	}
 
+	atlasDeployment := deploymentInAKO.GetCustomResource()
 	secrets := make([]string, 0)
 	for i := range databaseUsers.Items {
```
Comment (roothorp) [inline pkg/controller/atlasdeployment/advanced_deployment.go:122]: This loop doesn't seem to have test coverage - is it worth adding a database user to a test so we can test this?

**Review by roothorp — APPROVED** [review]
Couple of nit picks but looks good!

This is a huge PR but we have good coverage here (unit & e2e) so I'm pretty confident.

File: pkg/controller/atlasdeployment/advanced_deployment.go [hunk 4]
```diff
@@ -1,333 +1,128 @@
...
+		}
 
-	for _, d := range advancedDeployments.Results {
-		deploymentNames = append(deploymentNames, d.Name)
+		return r.ready(ctx, deploymentInAKO.GetCustomResource(), deploymentInAtlas)
+	case status.StateCREATING:
+		return r.inProgress(ctx, deploymentInAKO.GetCustomResource(), deploymentInAtlas, workflow.DeploymentCreating, "deployment is provisioning")
+	case status.StateUPDATING, status.StateREPAIRING:
+		return r.inProgress(ctx, deploymentInAKO.GetCustomResource(), deploymentInAtlas, workflow.DeploymentUpdating, "deployment is updating")
+	case status.StateDELETING, status.StateDELETED:
```
[the remark this answers is not in the document]
Comment (helderjs) [reply pkg/controller/atlasdeployment/advanced_deployment.go:106]: Until the moment we don't handle deleting state in any resource, it's basically fire and forget

File: pkg/controller/atlasdeployment/advanced_deployment.go [hunk 5]
```diff
@@ -1,333 +1,128 @@
...
-	deployment *akov2.AtlasDeployment,
-	atlasDeploymentAsAtlas *mongodbatlas.AdvancedCluster,
-) (*mongodbatlas.AdvancedCluster, workflow.Result) {
-	specDeployment, atlasDeployment, err := MergedAdvancedDeployment(*atlasDeploymentAsAtlas, *deployment.Spec.DeploymentSpec)
-	if err != nil {
-		return atlasDeploymentAsAtlas, workflow.Terminate(workflow.Internal, err.Error())
-	}
+	switch deploymentInAtlas.GetState() {
+	case status.StateIDLE:
+		if changes, occurred := deployment.ComputeChanges(deploymentInAKO, deploymentInAtlas); occurred {
```
[the remark this answers is not in the document]
Comment (helderjs) [reply pkg/controller/atlasdeployment/advanced_deployment.go:38]: While it looks like I'm also doing state machine here, I'm not. I just reorganized the code to remove duplication and make the flows (regular vs serverless) clearer.
I think this should be part when we actually refactor the controller in a state machine approach.

File: pkg/controller/atlasdeployment/advanced_deployment.go (hunk 3 above)
Comment (helderjs) [reply pkg/controller/atlasdeployment/advanced_deployment.go:122]: good point! the reason(s) I didn't add a test to cover these changes were two: 1) connection secrets are extensively tested on integration tests, which is a good safenet. 2) There a proposal to extract connection secret handling into a separate component or even a controller.

**Pull request merged**